ROS2 Polling Subscription Plugin

daggarwa opened the pull request:
Added polling subscription plugin for ROS 2 to the performance test. It can be used by specifying the argument `-c [ --communication ]` arg  Communication plugin to use as `ROS2PollingSub` when running performance test . 

To turn on the option for using Polling Subscription Plugin we need to compile the performance test with cmake args `PERFORMANCE_TEST_POLLING_SUBSCRIPTION_ENABLED` and value as `ON`  :

```bash
colcon build --cmake-clean-cache --cmake-args -DPERFORMANCE_TEST_POLLING_SUBSCRIPTION_ENABLED=ON
```
Sample usage for the plugin can be :
```bash
ros2 run performance_test perf_test -c ROS2PollingSubscription -l log -t Array16k --history_depth 100 --max_runtime 60 -p 1 -s 1 -r 20 --reliable --keep_last`
```
I have verified that this plugin builds and runs successfully on ApexOS version : `0d34c574917d`

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/apexai/performance_test/88)
<!-- Reviewable:end -->

monidzik commented:
@daggarwa you should put some information about this new option in readme file.

daggarwa commented:
> @daggarwa you should put some information about this new option in readme file.

@monidzik Sounds good! Will do

monidzik commented on performance_test/src/experiment_configuration/communication_mean.hpp:
```diff
@@ -40,6 +41,9 @@ inline std::ostream & operator<<(std::ostream & stream, const CommunicationMean
     return stream << "ROS2";
+  }
+  else if (cm == CommunicationMean::ROS2PollingSub) {
+    return stream <<"ROSPollingSub";
```
it should be also ROS2PollingSub

daggarwa replied:
@monidzik Sure I have incorporated this change. Thanks for pointing this out.

daggarwa replied:
Changed thanks!

monidzik commented on performance_test/src/communication_abstractions/ros2_communicator.hpp:
```diff
@@ -0,0 +1,186 @@
+  std::shared_ptr<::rclcpp::Publisher<DataType, std::allocator<void>>> m_publisher;
+};
+
+
```
you can do some clean-up in this file, there are couple of places where you have two unnecessary empty lines

daggarwa replied:
Removed the empty lines . Thanks

monidzik commented on performance_test/src/data_running/data_runner_factory.cpp:
```diff
@@ -20,6 +20,7 @@
 #include <memory>
 
 #include "../communication_abstractions/ros2_callback_communicator.hpp"
+#include "../communication_abstractions/ros2_waitset_communicator.hpp"
```
do you have to use relative paths instead of just #include "ros2_waitset_communicator.hpp"?

daggarwa replied:
Okay I was just being consistent with the already imported `ros2_callback_communicator.hpp`. Should I make this change?

monidzik replied:
I think it's not in the scope of this PR, I see it's in all the files done like this, so maybe we stay consistent.

daggarwa commented:
> I know you are working on correcting the qos settings so I can compile your branch, so let me check your code again when it's ready to run and CI is passing.

@monidzik I have made the requested changes (most of them except one : https://github.com/ApexAI/performance_test/pull/88#discussion_r328612243) need your input on that. 
Also the CI passes now. I believe the PR is ready for a second round of review. I have assigned it to you again. Thanks!

monidzik commented on performance_test/CMakeLists.txt:
```diff
@@ -155,6 +156,17 @@ set(sources
     src/utilities/statistics_tracker.hpp
 )
 
+#ROS2 Waitset
```
space missing between # and ROS2

daggarwa replied:
Sure thanks

monidzik commented on performance_test/CMakeLists.txt:
```diff
@@ -155,6 +156,17 @@ set(sources
 )
 
+#ROS2 Waitset
+option(PERFORMANCE_TEST_POLLING_SUBSCRIPTION_ENABLED "Enable Polling Subscription Plugin to be available only
```
and divide this line so it's not longer than 100chars.

daggarwa replied:
Corrected this.

monidzik commented on performance_test/src/communication_abstractions/ros2_callback_communicator.hpp:
```diff
@@ -21,165 +21,46 @@
   rclcpp::executors::SingleThreadedExecutor m_executor;
-  std::shared_ptr<::rclcpp::Publisher<DataType, std::allocator<void>>> m_publisher;
-  std::shared_ptr<::rclcpp::Subscription<DataType, std::allocator<void>>> m_subscription;
+  std::shared_ptr<::rclcpp::Subscription<DataType, std::allocator<void>>>
```
this can be in one line

daggarwa replied:
Corrected this. Thanks

monidzik commented on performance_test/src/communication_abstractions/ros2_waitset_communicator.hpp:
```diff
@@ -0,0 +1,77 @@
+        }
+      }
+    } catch (const rclcpp::TimeoutError &) {
+      std::cerr << "Waitset timed out without receiving a sample." << std::endl;
```
When I'm running the program I get this message just at the beginning, in an command line output:
---EXPERIMENT-START---
T_experiment,	T_loop,	received,	sent,	lost,	relative_loss,	....
Waitset timed out without receiving a sample.
1.047231,		1.047225,		20,	...

Is it a desired behavior?

daggarwa replied:
Yeah I think should be fine. Basically system will take time to stabilize initially (like the publishers get matched to subscribers etc) so can happen that it times out without receiving sample. @deeplearningrobotics Can also confirm if this indeed the desired behavior.

deeplearningrobotics replied:
@daggarwa, @monidzik: Yes, we should not wait infinitely here as then we can never terminate this thread. We just need to ensure that this error messages do not end up in the log file (which they don't as you print to `std::cerr`). So this should be fine.

daggarwa commented on performance_test/src/experiment_configuration/experiment_configuration.cpp:
```diff
@@ -74,7 +74,8 @@ void ExperimentConfiguration::setup(int argc, char ** argv)
     "The rate data should be published. Defaults to 1000 Hz. 0 means publish as fast as possible.")(
     "communication,c", po::value<std::string>()->required(),
-    "Communication plugin to use (ROS2, FastRTPS, ConnextDDSMicro, CycloneDDS)")("topic,t",
+    "Communication plugin to use (ROS2, FastRTPS, ConnextDDSMicro, CycloneDDS, ROS2PollingSub)")
```
Okay corrected that. Thanks!

daggarwa commented:
@monidzik I guess I have addressed all your review comments. Thanks so much for the review. As discussed I am now assigning the PR to @deeplearningrobotics for further review.

daggarwa commented:
> Ok, you still need to add the readme information with the option description, otherwise I have only couple of small remarks.

I forgot to respond on this earlier. I can do the update to the readme now and push that.

deeplearningrobotics commented on performance_test/src/communication_abstractions/ros2_waitset_communicator.hpp:
```diff
@@ -0,0 +1,77 @@
+    this->lock();
+    try {
+      rclcpp::Waitset<> m_waitset {m_polling_subscription};
+      auto wait_ret = m_waitset.wait(std::chrono::milliseconds(100));
```
Const correctness.

daggarwa replied:
Sure I will add that. Thanks

deeplearningrobotics commented on performance_test/src/communication_abstractions/ros2_waitset_communicator.hpp:
```diff
@@ -0,0 +1,77 @@
+    }
+    this->lock();
+    try {
+      rclcpp::Waitset<> m_waitset {m_polling_subscription};
```
This should be a member? You should not construct it every time.

daggarwa replied:
I had done that before. But `rclcpp::Waitset<>` doesn't have a default constructor.

deeplearningrobotics replied:
https://www.geeksforgeeks.org/when-do-we-use-initializer-list-in-c/ Paragraph 3

daggarwa replied:
@deeplearningrobotics If I change to the following :
```
  explicit ROS2WaitsetCommunicator(SpinLock & lock)
  : ROS2Communicator<Topic>(lock),
    m_polling_subscription(nullptr),
    m_waitset(m_polling_subscription){}

```
The  performance test will not work since we are initializing the `m_polling_subscription` which is needed by `m_waitset` here : https://github.com/ApexAI/performance_test/blob/3290-ros2-waitset/performance_test/src/communication_abstractions/ros2_waitset_communicator.hpp#L50 inside the `update_subscription` method. Is there any other workaround for this to get this to work?

daggarwa replied:
Okay made the changes accordingly

deeplearningrobotics commented on performance_test/src/communication_abstractions/ros2_communicator.hpp:
```diff
@@ -0,0 +1,178 @@
+    callback(*data);
+  }
+
+  void callback(DataType & data)
```
Const correctness.

daggarwa replied:
This callback calls `publish` https://github.com/ApexAI/performance_test/blob/3290-ros2-waitset/performance_test/src/communication_abstractions/ros2_communicator.hpp#L106  if the round trip mode is selected as `RELAY` which modifies the `Datatype data` so I can't put the Datatype data to be `const` here

deeplearningrobotics replied:
```


void publish(const DataType & data, const std::chrono::nanoseconds time) {
  DataType copy = data;
  publish(copy);
}

 void publish(DataType & data, const std::chrono::nanoseconds time)
  {
...}

template<class T>
void callback(T & data) {
static_assert(std::is_same<DataType, std::remove_cv<T>>::value);
}

```

daggarwa replied:
Okay made the changes accordingly

deeplearningrobotics commented on performance_test/src/communication_abstractions/ros2_waitset_communicator.hpp:
```diff
@@ -0,0 +1,77 @@
+  }
+
+private:
+  std::shared_ptr<::rclcpp::PollingSubscription<DataType, std::allocator<void>>>
```
Strange formatting, also you don't need `std::allocator...`. This should be a default parameter.

daggarwa replied:
Sure . Corrected the formatting ( even though `uncrustify` or `cpplint` did not show this). 
Remove `std::allocator` from the above

deeplearningrobotics replied:
I don't see this updated.

deeplearningrobotics commented on performance_test/src/communication_abstractions/ros2_communicator.hpp:
```diff
@@ -0,0 +1,178 @@
+  : Communicator(lock),
+    m_node(ResourceManager::get().ros2_node())
+  {
+    m_ROS2QOSAdapter = std::make_shared<rclcpp::QoS>(ROS2QOSAdapter(m_ec.qos()).get());
```
Why is that a pointer?

daggarwa replied:
Yes doesn't need to be a pointer. I can change that,

deeplearningrobotics commented on performance_test/src/communication_abstractions/ros2_communicator.hpp:
```diff
@@ -0,0 +1,178 @@
+      throw std::runtime_error("Unsupported QOS!");
+    }
+
+
```
Strange formatting, did you run uncrustify?

daggarwa replied:
Yes I ran uncrustify on all the files I have committed. I can run again to verify if there is any divergence

deeplearningrobotics commented on performance_test/src/communication_abstractions/ros2_callback_communicator.hpp:
```diff
@@ -21,164 +21,44 @@
-        std::bind(&ROS2CallbackCommunicator::callback, this, std::placeholders::_1), qos);
+      m_subscription = this->m_node->template create_subscription<DataType>(
+        Topic::topic_name() + this->m_ec.sub_topic_postfix(), *ros2QOSAdapter,
+        [this](const typename DataType::SharedPtr data) {return this->callback(data);});
```
You are returning `void` here, why?

daggarwa replied:
No need. Removing

deeplearningrobotics commented on performance_test/src/communication_abstractions/ros2_callback_communicator.hpp:
```diff
@@ -21,164 +21,44 @@
-  void update_subscription()
+  void update_subscription() override
   {
+    auto ros2QOSAdapter = this->m_ROS2QOSAdapter;
```
Why do you create this temporary here?

daggarwa replied:
Yes I can remove this. There is no need to do this temporary assignment here.

deeplearningrobotics commented on performance_test/src/communication_abstractions/ros2_communicator.hpp:
```diff
@@ -0,0 +1,178 @@
+   * \param qos The abstract QOS to derive the settings from.
+   */
+  explicit ROS2QOSAdapter(const QOSAbstraction qos)
+  : m_qos(qos)
```
Strange line break

daggarwa replied:
Okay I will correct that. `Uncrustify` or `cpplint` did not complain

daggarwa replied:
I had already run `uncrustify` when I committed this file . I tried making this change manually where I removed the line break but when I ran `uncrustify` again on this file, it introduced this line break again.

deeplearningrobotics commented on performance_test/src/communication_abstractions/ros2_communicator.hpp:
```diff
@@ -0,0 +1,178 @@
+
+protected:
+  std::shared_ptr<rclcpp::Node> m_node;
+  std::shared_ptr<rclcpp::QoS> m_ROS2QOSAdapter;
```
Why is this a shared ptr?

daggarwa replied:
No need . can be changed to simple object of type `rclcpp::QoS`

deeplearningrobotics commented on performance_test/src/experiment_configuration/experiment_configuration.cpp:
```diff
@@ -140,6 +141,13 @@ void ExperimentConfiguration::setup(int argc, char ** argv)
+      m_com_mean = CommunicationMean::ROS2PollingSub;
+#else
+      throw std::invalid_argument(
+              "You must compile with ApexOS to enable it as communication mean.");
```
Wrong error message. You must compile with POLLING SUB ENABLED

daggarwa replied:
Okay I will change the message appropriately.

deeplearningrobotics commented:
@monidzik: Please also review my comments.

daggarwa commented on performance_test/src/communication_abstractions/ros2_waitset_communicator.hpp:
```diff
@@ -0,0 +1,77 @@
+    try {
+      rclcpp::Waitset<> m_waitset {m_polling_subscription};
+      auto wait_ret = m_waitset.wait(std::chrono::milliseconds(100));
+      if (wait_ret.all()) {
```
Okay I will change that. Thanks

daggarwa commented on performance_test/src/experiment_configuration/communication_mean.hpp:
```diff
@@ -40,6 +43,10 @@ inline std::ostream & operator<<(std::ostream & stream, const CommunicationMean
   if (cm == CommunicationMean::ROS2) {
     return stream << "ROS2";
+#ifdef PERFORMANCE_TEST_POLLING_SUBSCRIPTION_ENABLED
+  } else if (cm == CommunicationMean::ROS2PollingSub) {
```
Okay I will change that. Thanks

daggarwa commented on performance_test/CMakeLists.txt:
```diff
@@ -155,6 +156,17 @@ set(sources
 )
 
+# ROS2 Waitset
+option(PERFORMANCE_TEST_POLLING_SUBSCRIPTION_ENABLED "Enable Polling Subscription Plugin to be
```
Okay

daggarwa commented on performance_test/src/communication_abstractions/ros2_waitset_communicator.hpp:
```diff
@@ -0,0 +1,77 @@
+    this->lock();
+    try {
+      rclcpp::Waitset<> m_waitset {m_polling_subscription};
+      auto wait_ret = m_waitset.wait(std::chrono::milliseconds(100));
```
Yes I will make that change.

daggarwa commented on performance_test/src/communication_abstractions/ros2_waitset_communicator.hpp:
```diff
@@ -0,0 +1,77 @@
+      auto wait_ret = m_waitset.wait(std::chrono::milliseconds(100));
+      if (wait_ret.all()) {
+        auto received_sample = m_polling_subscription->take();
+        DataType tmp = received_sample.data();
```
If I want to use a reference, I would have to use a const reference since `m_polling_subscription->take()` returns an object of type `LoanedMessageT` which is a typedef of `class LoanedMessage` which has the `data()` function returning a const ref of type `MessageT` . I cannot assign `tmp` on line 59 above as a const ref due to https://github.com/ApexAI/performance_test/pull/88#discussion_r329335925

deeplearningrobotics replied:
I think this is not fixed yet.

daggarwa commented:
> @monidzik: Please also review my comments.

@deeplearningrobotics @monidzik I have addressed's all of @deeplearningrobotics 's comments. Please review once. Thanks.

monidzik commented:
> > @monidzik: Please also review my comments.
> 
> @deeplearningrobotics @monidzik I have addressed's all of @deeplearningrobotics 's comments. Please review once. Thanks.

@daggarwa ok, next time please just assign to me, I didn't see this message, sorry!

monidzik commented on performance_test/CMakeLists.txt:
```diff
@@ -155,6 +156,17 @@ set(sources
 
+# ROS2 Waitset
+option(PERFORMANCE_TEST_POLLING_SUBSCRIPTION_ENABLED "Option to enable Polling Subscription Plugin.
+  The plugin can only work if ApexOS is present,otherwise it will fail." OFF)
```
present,otherwise -> space missing

daggarwa replied:
Okay added .

monidzik commented on performance_test/src/experiment_configuration/communication_mean.hpp:
```diff
@@ -40,6 +43,10 @@ inline std::ostream & operator<<(std::ostream & stream, const CommunicationMean
     return stream << "ROS2";
+#ifdef PERFORMANCE_TEST_POLLING_SUBSCRIPTION_ENABLED
+  } else if (cm == CommunicationMean::ROS2PollingSubscription) {
+    return stream << "ROS2PollingSub";
```
so here also ROS2PollingSubscription for consistency

daggarwa replied:
Okay changed this and also changed here as the name of the command line option : https://github.com/ApexAI/performance_test/blob/3290-ros2-waitset/performance_test/src/experiment_configuration/experiment_configuration.cpp#L145 and updated the description here :  https://github.com/ApexAI/performance_test/pull/88#issue-319478231 to reflect the same.

monidzik commented:
@deeplearningrobotics Divya has some comments for you to resolve, so leaving them for you!

daggarwa commented:
> > > @monidzik: Please also review my comments.
> > 
> > 
> > @deeplearningrobotics @monidzik I have addressed's all of @deeplearningrobotics 's comments. Please review once. Thanks.
> 
> @daggarwa ok, next time please just assign to me, I didn't see this message, sorry!

@monidzik Okay I will do that from next time. Thanks for reviewing this.

daggarwa commented:
@deeplearningrobotics  As discussed in the morning, I have committed changes to resolve   https://github.com/ApexAI/performance_test/pull/88#discussion_r329707857 and  https://github.com/ApexAI/performance_test/pull/88#discussion_r329718116

Can you please review the changes here : https://github.com/ApexAI/performance_test/pull/88/commits/aa11e1d46ca92f5ec3fd146a752e459cc7290e41
and continue your review of the PR. Thanks

deeplearningrobotics commented on performance_test/src/communication_abstractions/ros2_communicator.hpp:
```diff
@@ -0,0 +1,184 @@
+    m_publisher->publish(data);
+  }
+
+  void publish(const DataType & data, const std::chrono::nanoseconds time)
```
Documentation missing, also add some comment on why we need this.

daggarwa replied:
Okay done that .

deeplearningrobotics commented on performance_test/src/communication_abstractions/ros2_communicator.hpp:
```diff
@@ -0,0 +1,184 @@
+  {
+    static_assert(std::is_same<DataType,
+      typename std::remove_cv<typename std::remove_reference<T>::type>::type>::value,
+      "type mismatch");
```
Put a proper error message: "Parameter type passed to callback() does not match".

Be consistent with the rest of the code. See how nice of an error message is below. It is a proper english sentence.

daggarwa replied:
Sure I just changed it. I had put that message for testing initially.

daggarwa commented:
@deeplearningrobotics I think I have addressed all your comments now. Can you please review once. Assigning back to you.

deeplearningrobotics commented:
@monidzik: I reviewed the changes and the code looks good to me. Please test all the communication plugins and check if everything is still functional and then approve.

@daggarwa: FYI

deeplearningrobotics commented:
@daggarwa: Can you also run it trough pref to make sure we did not introduce performance bottlenecks.

dejanpan commented:
> @daggarwa: Can you also run it trough pref to make sure we did not introduce performance bottlenecks.

@deeplearningrobotics is "pref" a typo? Did you mean https://perf.wiki.kernel.org/index.php/Tutorial?

 It is possible that @daggarwa will not understand what is meant by running it through perf - can you point her to what explicitly should be done?

deeplearningrobotics commented:
> > @daggarwa: Can you also run it trough pref to make sure we did not introduce performance bottlenecks.
> 
> @deeplearningrobotics is "pref" a typo? Did you mean https://perf.wiki.kernel.org/index.php/Tutorial?
> 
> It is possible that @daggarwa will not understand what is meant by running it through perf - can you point her to what explicitly should be done?

@dejanpan: Yes, I mean `perf` sorry.

@daggarwa: Lets look together at the `perf` graphs when you are back at the office so I can show you what to look out for.

daggarwa commented:
> > > @daggarwa: Can you also run it trough pref to make sure we did not introduce performance bottlenecks.
> > 
> > 
> > @deeplearningrobotics is "pref" a typo? Did you mean https://perf.wiki.kernel.org/index.php/Tutorial?
> > It is possible that @daggarwa will not understand what is meant by running it through perf - can you point her to what explicitly should be done?
> 
> @dejanpan: Yes, I mean `perf` sorry.
> 
> @daggarwa: Lets look together at the `perf` graphs when you are back at the office so I can show you what to look out for.

@dejanpan Thank you for the clarifying question. 
@deeplearningrobotics Okay sounds good.

daggarwa commented:
> @daggarwa looks good to me, I checked also other communication means and everything works fine. The only thing is that you are still not mentioning anything about this option in readme - did you decide not to write there anything?

@monidzik I am planning to add `ROS2 Waitset` like here : https://github.com/ApexAI/performance_test/pull/88/commits/09fd145426e6cb3ca501d7ac9a4b0d84d30121e5 to be consistent with other implemented plugins

daggarwa commented:
> > @daggarwa looks good to me, I checked also other communication means and everything works fine. The only thing is that you are still not mentioning anything about this option in readme - did you decide not to write there anything?
> 
> @monidzik I am planning to add `ROS2 Waitset` like here : [09fd145](https://github.com/ApexAI/performance_test/commit/09fd145426e6cb3ca501d7ac9a4b0d84d30121e5) to be consistent with other implemented plugins

@monidzik I have added the changes above to the ReadMe.md. Kindly review.

daggarwa commented:
@deeplearningrobotics The profiling with perf tool is done. There are no performance impacts if we add this plugin in. Also , rebased the branch off master and cleaned up the commits. Please let me know if anything else is required before it gets merged in! Thanks
